processor/tailsamplingprocessor/internal/sampling/composite_test.go

```diff
@@ -0,0 +1,250 @@
+		}
+
+		expected := Sampled
+		if decision != expected {
+			t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+		}
+	}
+}
+
+func TestOnLateArrivingSpans_Composite(t *testing.T) {
+	n1 := NewNumericAttributeFilter(zap.NewNop(), "tag", 0, 100)
+	n2 := NewAlwaysSample(zap.NewNop())
+	timeProvider := &FakeTimeProvider{second: 0}
+	const totalSPS = 100
+	c := NewComposite(zap.NewNop(), totalSPS, []SubPolicyEvalParams{{n1, totalSPS / 2}, {n2, totalSPS / 2}}, timeProvider)
+	e := c.OnLateArrivingSpans(Sampled, nil)
+	assert.Nil(t, e)
+}
+
+func TestCompositeEvaluator2SubpolicyThrottling(t *testing.T) {
+
+	n1 := NewNumericAttributeFilter(zap.NewNop(), "tag", 0, 100)
+	n2 := NewAlwaysSample(zap.NewNop())
+	timeProvider := &FakeTimeProvider{second: 0}
+	const totalSPS = 100
+	c := NewComposite(zap.NewNop(), totalSPS, []SubPolicyEvalParams{{n1, totalSPS / 2}, {n2, totalSPS / 2}}, timeProvider)
+
+	trace := createTrace()
+
+	// We have 2 subpolicies, so each should initially get half the bandwidth
+
+	// First totalSPS/2 should be Sampled until we hit the rate limit
+	for i := 0; i < totalSPS/2; i++ {
+		decision, err := c.Evaluate(traceID, trace)
+		if err != nil {
+			t.Fatalf("Failed to evaluate composite policy: %v", err)
+		}
+
+		expected := Sampled
+		if decision != expected {
+			t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+		}
+	}
+
+	// Now we hit the rate limit for second subpolicy, so subsequent evaluations should result in NotSampled
+	for i := 0; i < totalSPS/2; i++ {
+		decision, err := c.Evaluate(traceID, trace)
+		if err != nil {
+			t.Fatalf("Failed to evaluate composite policy: %v", err)
+		}
+
+		expected := NotSampled
+		if decision != expected {
+			t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+		}
+	}
+
+	// Let the time advance by one second.
+	timeProvider.second++
+
+	// It is a new second, so we should start sampling again.
+	for i := 0; i < totalSPS/2; i++ {
+		decision, err := c.Evaluate(traceID, trace)
+		if err != nil {
+			t.Fatalf("Failed to evaluate composite policy: %v", err)
+		}
+
+		expected := Sampled
+		if decision != expected {
+			t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+		}
+	}
+
+	// Now let's hit the hard limit and exceed the total by a factor of 2
+	for i := 0; i < 2*totalSPS; i++ {
+		decision, err := c.Evaluate(traceID, trace)
+		if err != nil {
+			t.Fatalf("Failed to evaluate composite policy: %v", err)
+		}
+
+		expected := NotSampled
+		if decision != expected {
+			t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+		}
+	}
+
+	// Let the time advance by one second.
+	timeProvider.second++
+
+	// It is a new second, so we should start sampling again.
+	for i := 0; i < totalSPS/2; i++ {
```

Comment:
Same here -- why do you need this again?

Reply:
Confirming the behavior once we enter a new window. Again we should start sampling